### ISSUE TITLE
When will you release the code?

### ISSUE THREAD

**Issue by LLMforScience** [thu-ml/Noise-Contrastive-Alignment#1]
Thank you for your work. When will you release the code?

**Comment by ChenDRAG** [comment]
Thank you for your interest in our work! We plan to clean and release the codebase before publication. We may expedite this process if we find more demand for it. As our code is heavily based on the [alignment-handbook](https://github.com/huggingface/alignment-handbook/tree/8f6e5b666ba5ab3ee07d36e889d525386b1e75c9) repository, we highly recommend trying out their codebase for now. The pseudo code in the appendix provides every detail of NCA's reimplementation. Hyperparameters and other design choices are consistent with qlora settings in the [alignment-handbook](https://github.com/huggingface/alignment-handbook/tree/8f6e5b666ba5ab3ee07d36e889d525386b1e75c9).

**Comment by winglian** [comment]
@ChenDRAG would love to at least see the loss calculation code implementation for this.

**Comment by ChenDRAG** [comment]
The loss implementation under pairwise preference settings.

```
def preference_loss(chosen_pi_logps, chosen_ref_logps,
rejected_pi_logps, rejected_ref_logps,
beta, loss_type):
"""
chosen_pi_logps: policy logprobs for the preferred responses, shape (B, )
chosen_ref_logps: reference logprobs for the preferred responses, shape (B, )
rejected_pi_logps: policy logprobs for the dispreferred responses, shape (B, )
rejected_ref_logps: reference logprobs for the dispreferred responses, shape (B, )
beta: the parameterization coefficient that defines the residual model
loss_type: one of "InfoNCA", "NCA" or "DPO" loss
"""
chosen_rewards = (chosen_pi_logps - chosen_ref_logps) * beta # (B,)
rejected_rewards = (rejected_pi_logps - rejected_ref_logps) * beta # (B,)
if loss_type in ["DPO", "InfoNCA"]:
losses = -F.logsigmoid(chosen_rewards - rejected_rewards) # (B,)
elif loss_type == "NCA":
losses = - F.logsigmoid(chosen_rewards) \
- 0.5 * F.logsigmoid(-chosen_rewards) \
- 0.5 * F.logsigmoid(-rejected_rewards) # (B,)
return losses.mean()
```

The loss implementation under reward settings. (General case)

```
import torch.nn.functional as F
def reward_loss(pi_logps, ref_logps, rewards, alpha, beta, loss_type):
"""
pi_logps: policy logprobs for K responses, shape (B, K)
ref_logps: reference logprobs for K responses, shape (B, K)
rewards: reward labels for K responses, shape (B, K)
alpha: the reward temperature controlling strength of KL penalty
beta: the parameterization coefficient that defines the residual model
loss_type: could be either "InfoNCA" or "NCA" loss
"""
soft_labels = (rewards / alpha).softmax(dim=-1) # (B, K)
model_rewards = (pi_logps - ref_logps) * beta # (B, K)
if loss_type == "InfoNCA":
model_logps = model_rewards.log_softmax(dim=-1) # (B, K)
losses = - (soft_labels * model_logps).sum(dim=-1) # (B,)
elif loss_type == "NCA":
optimization = - (soft_labels * F.logsigmoid(model_rewards)).sum(dim=-1) # (B,)
regularization = - F.logsigmoid(-model_rewards).mean(dim=-1) # (B,)
losses = optimization + regularization # (B,)
return losses.mean()
```

Hope this could help. @winglian
